### ISSUE TITLE
`invalid-type-form` emitted for unusual index operations in non-type-expression contexts

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#727]
ty emits `invalid-type-form` for non-type expressions like the following that do not produce errors at runtime:
```py
tuple[1]
```

While the above case may seem nonsensical, there are real-world code examples where a strict treatment of non-type `tuple[ ]` expressions under the restricted grammar of type-expressions would lead to false positives (https://github.com/astral-sh/ruff/pull/18991#issuecomment-3018171342). For example:
```py
runtime_type_representation = tuple[(int, str) * 3]
```

**Comment by AlexWaygood** [comment]
One way to fix this might be to:
1. Refactor `infer_type_expression()` to return a `Result` rather than eagerly emitting diagnostics (this might also help with https://github.com/astral-sh/ty/issues/667)
2. In non-annotation contexts, attempt to parse the inner value as a type expression (and infer a `Type::GenericAlias` type if so, like we do on `main`), but if that returns an `Err`, fall back to `Type::NominalInstance("GenericAlias")` without emitting a diagnostic

**Comment by AlexWaygood** [comment]
#2905 notes that this problem can also come up in (uncommon) cases involving `pydantic` and `typing.Literal`, for example:

```py
from typing import Literal
import pydantic

my_string = "hello world!"

my_dynamic_class = pydantic.create_model(
    "my_dynamic_class",
    literal_string=Literal[
        my_string
    ],  # This gives an invalid-type-form error on Literal[my_string]
)

print(my_dynamic_class(literal_string="hello world!").literal_string) # Prints "hello world!" without error
```

**Comment by dsfaccini** [comment]
> This comment was posted by Claude Code using claude-fable-5 on behalf of David.

@AlexWaygood asked on #2905 whether patterns like this occur regularly in real codebases. Datapoint: yes, in [pydantic-ai](https://github.com/pydantic/pydantic-ai) — 5 diagnostics on ty 0.0.72, across three modules, all of them runtime construction of pydantic models whose shape is only known at runtime.

Two additions to what's already on this issue:

**1. `Annotated[...]` hits it too**, not just `tuple[...]` and `Literal[...]`:

```py
from typing import Annotated, Literal


class Tag:
    def __init__(self, name: str) -> None: ...


def build_schema(name: str, member: type) -> tuple[object, object]:
    literal_type = Literal[name]
    tagged_type = Annotated[member, Tag(name)]
    return literal_type, tagged_type
```

```console
$ uvx ty check repro.py
repro.py:9:28: error[invalid-type-form] Type arguments for `Literal` must be `None`, a literal value (int, bool, str, or bytes), or an enum member
repro.py:10:29: error[invalid-type-form] Variable of type `type` is not allowed in a type expression
Found 2 diagnostics
```

pyright is clean on this file; mypy reports both.

**2. The real-world sites**, for what the pattern looks like in practice:

- [`native_tools/__init__.py`](https://github.com/pydantic/pydantic-ai/blob/main/pydantic_ai_slim/pydantic_ai/native_tools/__init__.py) — inside `__get_pydantic_core_schema__`, building a discriminated union over a registry populated by `__init_subclass__`: `Annotated[tool, pydantic.Tag(tool.kind)]` for each member, then `Annotated[Union[tuple(tools_annotated)], pydantic.Discriminator(...)]`. The set of members isn't statically known — third-party packages register into it.
- [`_spec.py`](https://github.com/pydantic/pydantic-ai/blob/main/pydantic_ai_slim/pydantic_ai/_spec.py) — `Literal[name]` where `name` is a tool name read from a signature, plus a functional `TypedDict(f'{prefix}_{name}', fields)` built from that signature.
- [`agent/spec.py`](https://github.com/pydantic/pydantic-ai/blob/main/pydantic_ai_slim/pydantic_ai/agent/spec.py) — `list[Union[tuple(capability_schema_types)]]`, same registry idea.

The common thread is that all of these feed a value into pydantic's `handler(...)` / `create_model(...)` / annotation slot at runtime — the expression is a value being passed to a library, and nothing downstream treats it as a static annotation.

Notes: ty 0.0.72 via `uvx ty check`.

**Comment by carljm** [comment]
#1711 is the "full" approach to fixing this (but we can probably also fix it in a more targeted way)
